### PR TITLE
Multithreaded Batch_images generator

### PR DESCRIPTION
**Pull request by ereish64**
The build_batch option when set to true destroys previously generated images, even if those images can be used again. I have added an if statement that checks to see if the batch directory for the dataset exists, if one doesn't then the computer creates the directory and generates new batch images. Otherwise, the computer will use the existing images in the directory.

However, there are two limitations to this method. If the computer is interrupted while it is building the batch images or the dataset is altered after the batch images have already been built, the batch image directory for that dataset must be deleted. If the directory is not deleted, it will result in a index error during training.

While this is not an ideal fix, it still saves quite a bit of time if training the same model with no modifications to dataset.

### CODE REVIEW

**Comment by jiny2001** [comment]
Hi ereish64, 
Thank you for sending the request! Sorry for replying late.

Actually, it 's not just destroying and building the batch images agin. Please check this is_batch_exist() function which you removed from the code.

https://github.com/jiny2001/dcscn-super-resolution/blob/master/helper/loader.py#L196

It checks if there is a directory and also checks if the generated batch images has same image size and so on. I guess current code does a bit more than your code. Thx!!

**Comment by ereish64** [comment]
That's odd. It didn't seem to be working properly on my computer. Perhaps I
did something wrong?

Anyway, I'm working on a multithreaded version of the build_batch command
that should be done at some point. I know your not supposed to use the tool
on large datasets, but it's still pretty slow with only one thread so I
thought I might try to improve that.

**Comment by jiny2001** [comment]
Humm..

Please check if there is a batch_images.ini file. The file will be generated after it finished building batch images. It did work in my environment.

Anyway, multi threaded version of build_batch sounds awesome! Please let me know when you have done it. Thx!  :)

**Comment by ereish64** [comment]
One more thing, I seem to be getting a lot of noise when I train my own
model. Any idea what would cause this?

Thanks again

**Comment by jiny2001** [comment]
A noise... What type of noise? You mean the inference result gets noise. Correct?

I have been using the code with many network architecture and training images but have never seen a noise. Could you share your sample result here?

**Comment by ereish64** [comment]
My computer's power supply actually gave out yesterday, so I can't access
the photos at the moment. However, I found a picture on the internet that
represents it quite well.

Perhaps I wasn't clear, I'm talking about image noise. I would input an
image like the photo on the right side and end up with a picture more like
the photo on the left. It isn't quite that bad, but it is noticable.

I'm using 200 ish high resoloution images for training, each image has
already been converted to a grayscale bitmap with the provided convert_y
script.

Any ideas? I'm kind of stumped.

**Comment by jiny2001** [comment]
Hi, could you share the URL of the photo you mentioned? :)

**Comment by ereish64** [comment]
https://www.google.com/search?q=image+noise&safe=strict&client=ms-android-blu&prmd=isnv&source=lnms&tbm=isch&sa=X&ved=0ahUKEwi_nMKZsY3kAhUFnlkKHU23DVAQ_AUIGCgB#imgrc=hsOtGc3oWEG_cM

**Comment by jiny2001** [comment]
Oh, OK. That's noisy...

Have you checked if your pre-built batch images don't have that kind of noise?
If not one big possibility came up to my mind is, lack of training iteration. How much number of steps(images) do you try for each epoch?

**Comment by ereish64** [comment]
How many epochs did you train the example network on?

**Comment by jiny2001** [comment]
Hi, it really depend on the model size and input batch image size.

But for --training_images, usually I use 20,000 for smaller model to 100,000 for normal or larger model.

**Comment by ereish64** [comment]
But how many epochs was the model trained on? I'll have my computer back
and running by tomorrow so the multithreaded code should be done by the end
of this week. Thanks for helping me out, I really appreciate it. I'm 17 and
I'm still a little new to super resoloution neural networks so it's nice to
have someone who knows a lot about this kind of stuff.

**Comment by jiny2001** [comment]
Hi, that's amazing you are so young!! I'm happy to work for you. :) 
By default, the model will be trained with 9 epochs for the first and then make the learning rate half.
Here is the list of learning rates for the model.

0.002, 0.001, 5e-4, 2.5e-4, 1.25e-4, 6.25e-5, 3.12e-5

So there are 7 steps and each step has 9 epochs so the total epoch will be 7 x 9 = 63 epochs.
But in my experience, you can try 10,000 images for 1 epoch and also you can finish the training when the learning rate reaches 1.25e-4. In this case the training time would get approximately 1/4.

**Comment by ereish64** [comment]
as promised, the multi-threaded code is done. Let me know if i missed something. The file naming for the output batch files is pretty sloppy, but that's the best way i could come up with, feel free to change it as you wish

**Comment by ereish64** [comment]
scratch that, the loader is unable to load the batch files because of the way the threads name them. I'm correcting the issue and will also multithread the image loader.

**Comment by ereish64** [comment]
I defiantly wouldn't use this pull request yet, i still have some bugs to work out. For example, each process is trying to re-import all of the libraries that the main program uses, even if it doesnt use them. 8 numpy and 8 Scipy instances slams RAM usage. I think that issue is under the rug now, but i still need to double check

**Comment by ereish64** [comment]
@jiny2001 The multi threaded code is done.
